### PR TITLE
Revenge of csv

### PR DESCRIPTION
**Pull request by mgrauer**
@essamjoubori 

Can you take over on this branch?  It contains much of what you put in earlier.

There may be some style errors, and can you fix the issue where it isn't finding a couple image resources (you'll see in the browser console when you upload/display a csv file), but it should be fairly close, and I can stop bugging you soon asking you about CSV!

Do you want to try work out the issues with the infinite scroll feature of datatables, or just punt for now?

### CODE REVIEW

**Comment by mgrauer** [comment]
@kotfic recently made the very reasonable request that PRs describe what features they are bringing and why.

This feature adds a CSV viewer widget for datasets with mimetype of `text/csv`.  It will display a table icon on the dataset panel

screen shot 2016-02-16 at 11 15 12 pm

which when clicked will bring up the dataset in a CSV viewer

screen shot 2016-02-16 at 11 15 22 pm

**Comment by aashish24** [comment]
The widget looks awesome @mgrauer

**Comment by essamjoubori** [comment]
@mgrauer I will take over this branch.

**Comment by mgrauer** [comment]
I'm posting here only because it relates to CSV, not to say we should use it, and certainly not now.

I saw this [jquery sheet](http://spreadsheets.github.io/WickedGrid/) (maybe now named wicked grid?), that seemed like a full spreadsheet in the browser.  It did have a php component, and I performed the most cursory of perusals, but it could be worth looking into more.

**Comment by kotfic** [comment]
@mgrauer thanks for the PR tagline,  this is very helpful!

On Wed, Feb 17, 2016 at 9:54 AM, mgrauer <notifications@github.com> wrote:

> I'm posting here only because it relates to CSV, not to say we should use
> it, and certainly not now.
>
> I saw this jquery sheet <http://spreadsheets.github.io/WickedGrid/>
> (maybe now named wicked grid?), that seemed like a full spreadsheet in the
> browser. It did have a php component, and I performed the most cursory of
> perusals, but it could be worth looking into more.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Kitware/minerva/pull/312#issuecomment-185238449>.
>



-- 
Christopher Kotfila
R&D Engineer
Kitware Inc.
http://www.kitware.com/company/team/kotfila.html

**Comment by essamjoubori** [comment]
@mgrauer there were some issues with datatables npm packages so I substituted them with a customized version from their [download builder](https://datatables.net/download/index). 

Also, I was able to get the infinite scrolling to work. 

What needs to be fixed: 
  - Headers are not aligned with table body on initial view, but once you start scrolling the headers will align perfectly. This seems to be a known issue. Kate might be able to fix that with some css magic. 

infinite_scroll

**Comment by mgrauer** [comment]
@essamjoubori 

I won't be able to QA this until tomorrow, but could you remove the cancel/save buttons at the bottom of the widget?  Or set them as showable based on a param to the widget, and for this read only case pass in a param so that they don't display.

Awesome animated gif, and way to go with the infinite scroll!

:zap:

**Comment by aashish24** [comment]
> Awesome animated gif, and way to go with the infinite scroll!

+1 @essamjoubori. How did you generate this gif?

**Comment by mgrauer** [comment]
@essamjoubori I'm getting an error with a csv when I try to scroll down, see the screenshot below.  The csv you sent me works fine.  The one I tried only had 25 rows, I'll email it to you in case there is something strange about it.  

Initially only 15 rows are displayed, perhaps because my test file doesn't have a full extra page of rows (only 10 more instead of 15) it fails when I scroll down and attempt to load more?

screen shot 2016-02-23 at 7 10 28 pm

**Comment by essamjoubori** [comment]
@mgrauer I fixed the bug with the infinite scrolling. Ready for QA.

**Comment by mgrauer** [comment]
Getting closer!

(1) Something simple, can we have have 40 rows displayed by default?

(2) I noticed that the reports of row numbers are off.

When I'm in the middle of the big csv, it says 106 to 106 of whatever, but it should say 96 to 106 of whatever.

screen shot 2016-02-24 at 10 37 18 am

At the start of the small csv, it says 1 to 0 of 25, but it should say 1 to 10 of 25.

screen shot 2016-02-24 at 10 37 30 am

When I get to the end of the 25, it says 15 to 15 of 25, but it should say 15 to 25 of 25.

screen shot 2016-02-24 at 10 38 38 am

(3) A problem I saw is that it wasn't loading the last line of the small CSV.  For me the last line displayed in the browser is "Sao Paulo", but in the csv file the last line is "Tocantins".

File: web_external/js/models/DatasetModel.js [hunk 1]
```diff
@@ -91,7 +99,7 @@ minerva.models.DatasetModel = minerva.models.MinervaModel.extend({
                     type: 'wms'
                 };
             } else {
-                console.error('Unknown dataset_type for geo_render purposes: ' + mm.dataset_type);
+                console.log('Unknown dataset_type for geo_render purposes: ' + mm.dataset_type);
```
Comment (danlamanna) [inline web_external/js/models/DatasetModel.js]: Out of curiosity, why this change?

Comment (mgrauer) [inline web_external/js/models/DatasetModel.js]: It's fine to have a csv file that can display as a csv, but currently there is no mechanism to get this displayed on the map.  We don't display the eyeball to display it on the map, so there isn't a problem of trying to render something that can't be rendered on the map.

Hence, it's not an error, but it might be something to be aware of.  Probably it could be removed altogether, but at this stage I'm erring on the side of letting developers know what is going on.

**Comment by mgrauer** [comment]
Todo for @mgrauer: de-namespace events, based on the conclusion of #319 .

**Comment by mgrauer** [comment]
>Todo for @mgrauer: de-namespace events, based on the conclusion of #319 .

Done.

@essamjoubori  It's all yours.  Note that you will have to delete all the items in your Girder user's minerva/session folder since the storage format of sessions has now changed.

**Comment by essamjoubori** [comment]
@mgrauer I fixed the mentioned bugs above. Re: the table headers misalignment, let's open an issue and then I will work on it on a separate PR.

**Comment by mgrauer** [comment]
Great, thanks @essamjoubori .  I added a few cleanup tasks in #323.

LGTM.  Merging.  :fireworks:
